File: boi/beetle/submissions/partially_accepted/gemini_pareto_front.py
import sys

def solve():
    # Read all tokens from standard input
    input_data = sys.stdin.read().split()
    if not input_data:
        return
    
    n = int(input_data[0])
    m = int(input_data[1])
    
    drops_input = []
    has_zero = False
    for i in range(n):
        x = int(input_data[2+i])
        drops_input.append(x)
        if x == 0:
            has_zero = True
            
    # Add artificial starting point if there isn't a drop at 0
    if not has_zero:
        drops_input.append(0)
        
    drops_input.sort()
    
    N = len(drops_input)
    X = drops_input
    
    # Establish original quantities of water at each index
    M = []
    Z = -1
    for i in range(N):
        if X[i] == 0:
            Z = i
            M.append(m if has_zero else 0)
        else:
            M.append(m)
            
    # dp[l][r] will store a tuple: (pareto_front_at_l, pareto_front_at_r)
    # where each pareto front is a list of (time, water) pairs.
    dp = [[([], []) for _ in range(N)] for __ in range(N)]
    
    # Base Case at the origin 0
    dp[Z][Z] = ([(0, M[Z])], [(0, M[Z])])
    
    # Track global maximum possible water that can be consumed
    max_ans = M[Z]
    
    # Bottom-Up Dynamic Programming (by length of interval L)
    for L in range(2, N + 1):
        # Determine valid intervals [l, r] that MUST contain the starting coordinate Z
        for l in range(max(0, Z - L + 1), min(Z + 1, N - L + 1)):
            r = l + L - 1
            
            # 1. Compute pos = 0 (we just arrived at `l`)
            pts0 = []
            if l + 1 <= Z:
                # Arrived at `l` from `l+1`
                d1 = X[l+1] - X[l]
                if dp[l+1][r][0]:
                    pts0.extend([(t + d1, w + (M[l] - t - d1 if M[l] > t + d1 else 0)) for t, w in dp[l+1][r][0]])
                
                # Arrived at `l` from `r`
                d2 = X[r] - X[l]
                if dp[l+1][r][1]:
                    pts0.extend([(t + d2, w + (M[l] - t - d2 if M[l] > t + d2 else 0)) for t, w in dp[l+1][r][1]])
                    
            pareto0 = []
            if pts0:
                # Merge into Pareto front: sort by shortest time first, then largest water to break ties
                pts0.sort(key=lambda x: (x[0], -x[1]))
                best_water = -1
                for t, w in pts0:
                    if w > best_water:
                        pareto0.append((t, w))
                        best_water = w
                        if w > max_ans:
                            max_ans = w
                            
            # 2. Compute pos = 1 (we just arrived at `r`)
            pts1 = []
            if r - 1 >= Z:
                # Arrived at `r` from `l`
                d1 = X[r] - X[l]
                if dp[l][r-1][0]:
                    pts1.extend([(t + d1, w + (M[r] - t - d1 if M[r] > t + d1 else 0)) for t, w in dp[l][r-1][0]])
                
                # Arrived at `r` from `r-1`
                d2 = X[r] - X[r-1]
                if dp[l][r-1][1]:
                    pts1.extend([(t + d2, w + (M[r] - t - d2 if M[r] > t + d2 else 0)) for t, w in dp[l][r-1][1]])
                    
            pareto1 = []
            if pts1:
                # Merge into Pareto front: sort by shortest time first, then largest water to break ties
                pts1.sort(key=lambda x: (x[0], -x[1]))
                best_water = -1
                for t, w in pts1:
                    if w > best_water:
                        pareto1.append((t, w))
                        best_water = w
                        if w > max_ans:
                            max_ans = w
                            
            dp[l][r] = (pareto0, pareto1)

    # tot_size = 0
    # for l in range(N):
    #     for r in range(l, N):
    #         tot_size += len(dp[l][r][0]) + len(dp[l][r][1])
    # print("Total size of all pareto fronts:", tot_size)
    # bins = N
    # bin_values = [0] * bins
    # for l in range(N):
    #     for r in range(l, N):
    #         toadd = len(dp[l][r][0]) + len(dp[l][r][1])
    #         bin_values[l] += toadd
    # print("Bin distribution of pareto front sizes:", bin_values)

    print(max_ans)

if __name__ == '__main__':
    solve()

File: boi/beetle/submissions/accepted/gemini_pareto_front.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// Structure to hold our Pareto front configurations
struct Point {
    int t; // time elapsed
    int w; // water collected
};

// DP State: dp[length_parity][left_index][pos]
// pos = 0 means we are at the left bound 'l'
// pos = 1 means we are at the right bound 'r'
vector<Point> dp[2][310][2];

// Merges two sorted Pareto fronts while calculating distances and dropping sub-optimal configurations
inline void merge_and_pareto(const vector<Point>& src1, int d1, int val, 
                             const vector<Point>& src2, int d2, 
                             vector<Point>& out, int& global_max) {
    int i = 0, j = 0;
    int n1 = src1.size(), n2 = src2.size();
    out.clear();
    
    int best_w = -1;

    // Two-pointer merge (since sources are already strictly increasing in time)
    while (i < n1 || j < n2) {
        int t_new, w_old;
        
        if (i < n1 && j < n2) {
            int t1 = src1[i].t + d1;
            int t2 = src2[j].t + d2;
            if (t1 < t2) {
                t_new = t1;
                w_old = src1[i].w;
                i++;
            } else if (t2 < t1) {
                t_new = t2;
                w_old = src2[j].w;
                j++;
            } else { // Times are equal, pick the configuration that had more water
                t_new = t1;
                w_old = max(src1[i].w, src2[j].w);
                i++;
                j++;
            }
        } else if (i < n1) {
            t_new = src1[i].t + d1;
            w_old = src1[i].w;
            i++;
        } else {
            t_new = src2[j].t + d2;
            w_old = src2[j].w;
            j++;
        }
        
        // Add water gained at the newly visited drop
        int w_new = w_old + (val > t_new ? val - t_new : 0);
        
        // If it strictly improves the maximum water for this specific time frame, keep it
        if (w_new > best_w) {
            out.push_back({t_new, w_new});
            best_w = w_new;
            if (w_new > global_max) {
                global_max = w_new;
            }
        }
    }
}

int main() {
    // Fast I/O
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int n, m_val;
    if (!(cin >> n >> m_val)) return 0;
    
    vector<int> X;
    X.reserve(n + 1);
    
    bool has_zero = false;
    for (int i = 0; i < n; i++) {
        int x;
        cin >> x;
        X.push_back(x);
        if (x == 0) has_zero = true;
    }
    
    // Add artificial starting point if there isn't a drop at 0
    if (!has_zero) {
        X.push_back(0);
    }
    
    sort(X.begin(), X.end());
    
    int N = X.size();
    vector<int> M(N);
    int Z = -1;
    
    // Establish original quantities of water at each index
    for (int i = 0; i < N; i++) {
        if (X[i] == 0) {
            Z = i;
            M[i] = has_zero ? m_val : 0;
        } else {
            M[i] = m_val;
        }
    }
    
    int global_max = M[Z];
    
    // Base Case initialization at Length L = 1
    int prev = 1 % 2;
    dp[prev][Z][0].push_back({0, M[Z]});
    dp[prev][Z][1].push_back({0, M[Z]});
    
    // Bottom-Up Dynamic Programming over window length L
    for (int L = 2; L <= N; L++) {
        int curr = L & 1;
        prev = curr ^ 1;
        
        // Determine valid bounds for the interval [l, r] that MUST encapsulate Z
        int l_start = max(0, Z - L + 1);
        int l_end = min(Z, N - L);
        
        for (int l = l_start; l <= l_end; l++) {
            int r = l + L - 1;
            
            // 1. Compute pos = 0 (we just arrived at `l`)
            dp[curr][l][0].clear();
            if (l + 1 <= Z) {
                int d1 = X[l+1] - X[l];
                int d2 = X[r] - X[l];
                merge_and_pareto(dp[prev][l+1][0], d1, M[l],
                                 dp[prev][l+1][1], d2,
                                 dp[curr][l][0], global_max);
            }
            
            // 2. Compute pos = 1 (we just arrived at `r`)
            dp[curr][l][1].clear();
            if (r - 1 >= Z) {
                int d1 = X[r] - X[l];
                int d2 = X[r] - X[r-1];
                merge_and_pareto(dp[prev][l][0], d1, M[r],
                                 dp[prev][l][1], d2,
                                 dp[curr][l][1], global_max);
            }
        }
    }
    
    cout << global_max << "\n";
    
    return 0;
}